### PR TITLE
Link libdl as interface library

### PR DESCRIPTION
**Pull request by masterleinad**
Fixes https://github.com/kokkos/kokkos/issues/4824. Basically, this just uses `-ldl` instead of the full library path for linking but we still try to find the library and add the include path for `dlfcn.h`.
An alternative to changing `${NAME}_FOUND_LIBRARIES` for interface libraries, would be to not provide the library name in the initial call but add `${CMAKE_DL_LIBS}` via `target_link_libraries` later.
I opted for the current approach since it feels less of a workaround, and interface libraries are not supposed to depend on hardcoded library paths anyway.

### CODE REVIEW

**Review by nmm0 — APPROVED** [review]
Looks good to me. I think it's better to link the library like you did rather than adding yet another code path in the cmake

**Review by cz4rs — APPROVED** [review]

File: cmake/Modules/FindTPLLIBDL.cmake [hunk 1]
```diff
@@ -1 +1 @@
-KOKKOS_FIND_IMPORTED(LIBDL HEADER dlfcn.h  LIBRARY dl)
+KOKKOS_FIND_IMPORTED(LIBDL HEADER dlfcn.h INTERFACE LIBRARIES ${CMAKE_DL_LIBS})
```
[the remark this answers is not in the document]
Comment (masterleinad) [reply cmake/Modules/FindTPLLIBDL.cmake:1]: This required fewer changes. `KOKKOS_CREATE_IMPORTED_TPL` (called by `KOKKOS_FIND_IMPORTED`) treats `LIBRARY`(referenced as `TPL_LIBRARY`) as `IMPORTED_LOCATION` but uses `TARGET_LINK_LIBRARIES` for `LIBRARIES` (referenced as `TPL_LINK_LIBRARIES`). In particular, we currently error out if `KOKKOS_CREATE_IMPORTED_TPL` called for an interface library gets a `LIBRARY` argument passed.

**Comment by cz4rs** [comment]
> This required fewer changes. `KOKKOS_CREATE_IMPORTED_TPL` (called by `KOKKOS_FIND_IMPORTED`) treats `LIBRARY`(referenced as `TPL_LIBRARY`) as `IMPORTED_LOCATION` but uses `TARGET_LINK_LIBRARIES` for `LIBRARIES` (referenced as `TPL_LINK_LIBRARIES`). In particular, we currently error out if `KOKKOS_CREATE_IMPORTED_TPL` called for an interface library gets a `LIBRARY` argument passed.

Ok, so the difference between passing `LIBRARY` and `LIBRARIES` is somewhat surprising. The documentation for the macro doesn't really help:
https://github.com/kokkos/kokkos/blob/86af51d754810ddeb342fad2ab740143dfca5872/cmake/kokkos_functions.cmake#L632-L638

And if I didn't look at the implementation, I would intuitively expect something like:
```CMake
  FOREACH(LIB ${TPL_LIBRARIES})
    do-the-same-stuff-we-do-for-single-library()
  ENDFOREACH()
```

**Review by cz4rs — APPROVED** [review]
Looks good to me!

**Pull request merged**